examples/practical-coalton/1_simple_database.lisp

```diff
@@ -0,0 +1,250 @@
+    #:coalton
+    #:coalton-prelude)
+  (:import-from
+   #:alexandria
+   :with-gensyms)
+  (:local-nicknames
+   (#:vec #:coalton-library/vector)
+   (#:lst #:coalton-library/list)
+   (#:cel #:coalton-library/cell)
+   (#:str #:coalton-library/string)))
+(in-package :practical-coalton.simple-database)
+
+(named-readtables:in-readtable coalton:coalton)
+
+;;; Port of Chapter 3 from Practical Common Lisp, by Peter Seibel,
+;;; to Coalton.
+;;;
+;;; In PCL, the database is stored in a global variable. The functions
+;;; modifying the database write to that global. This is possible in
+;;; Coalton (by defining a CELL global variable), and we take that approach
+;;; in the port of chapter 23. In this chapter, we define a mutable Database
+;;; type, and each of the functions that read or write to a Database take
+;;; it as an argument.
+;;;
+;;; Note that the CL macros generating Coalton functions work very well
+;;; with the type system. Even though the macro code itself is not type
+;;; checked, the code it generates is. This way the type system will catch
+;;; any type errors caused by incorrect macro code or incorrect usage of
+;;; a macro.
+
+;;;
+;;; Define the types required to create a mutable database and basic
+;;; functions to work with them.
+;;;
+
+(coalton-toplevel
+  (define-struct CD
+    (title String)
+    (artist String)
+    (rating Integer)
+    (ripped Boolean))
+
+  (repr :transparent)
+  (define-type Database
+    (Database (Cell (List CD))))
+
+  (declare cd-data (Database -> (Cell (List CD))))
+  (define (cd-data (Database cd-cell))
+    cd-cell)
+
+  (declare from-cds ((List CD) -> Database))
+  (define (from-cds cd-list)
+    (Database (cel:new cd-list)))
+
+  (declare cds (Database -> (List CD)))
+  (define (cds (Database cds-data))
+    (cel:read cds-data))
+
+  (declare new-database (Unit -> Database))
+  (define (new-database)
+    (Database (cel:new Nil)))
+
+  (declare add-record (Database -> CD -> Database))
+  (define (add-record db cd)
+    (cel:push! (cd-data db) cd)
+    db)
+
+  (declare dump-cd (CD -> Unit))
+  (define (dump-cd cd)
+    (traceobject "title" (.title cd))
+    (traceobject "artist" (.artist cd))
+    (traceobject "rating" (.rating cd))
+    (traceobject "ripped" (.ripped cd)))
+
+  (declare dump-db (Database -> Unit))
+  (define (dump-db (Database cd-data))
+    (for cd in (cel:read cd-data)
+        (dump-cd cd)
+        (trace "--"))
+    Unit))
+
+;;;
+;;; Query DSL implemented as Common Lisp macros that generate Coalton code.
+;;;
+
+(cl:defun keyword-to-struct-accessor (keyword)
+  "Converts a keyword to a Coalton struct accessor.
+
+This is useful for converting keywords to Coalton struct accessors in the
+query DSL, like:
+
+(where :rating 8), in which :rating will eventually be transformed into
+(.rating cd) in the generated Coalton code.
+
+Example:
+
+(keyword-to-struct-accessor :title) => .TITLE"
+  (cl:intern (cl:concatenate 'cl:string (cl:string '#:.) (cl:string keyword))))
```

Comment:
(concatenate 'string "." (string keyword)) works

Reply:
I tried that and it didn't work. I think it shouldn't work because these are the CL versions of those functions?